Adds support to edit bucket class with namespace store resources

vbnrh opened the pull request:
https://marvelapp.com/prototype/fa616je/screen/74338415

cloudbehl commented:
creation of namespace store breaking.
```
Uncaught TypeError: Cannot read property 'items' of null
    at NamespaceStoreDropdown (namespace-store-dropdown.tsx:37)
```

cloudbehl commented:
The namespace store is not present in my system and the page is breaking and that should not be the case. Need to have valid messaging.

afreen23 commented on frontend/packages/ceph-storage-plugin/src/components/bucket-class/modals/edit-backingstore-modal.tsx:
```diff
@@ -54,34 +106,110 @@ const BucketClassEditModal = withHandlePromise<
+      }
     }
+    if (isNamespaceType) {
+      if (state.namespacePolicyType === NamespacePolicyType.SINGLE) {
+        bucketClass.spec.namespacePolicy.single.resource = getName(state.readNamespaceStore[0]);
+      }
+      if (state.namespacePolicyType === NamespacePolicyType.MULTI) {
+        bucketClass.spec.namespacePolicy.multi.writeResource = getName(
+          state.writeNamespaceStore[0],
+        );
+        bucketClass.spec.namespacePolicy.multi.readResources = state.readNamespaceStore.map(
+          getName,
+        );
+      }
+
+      if (state.namespacePolicyType === NamespacePolicyType.CACHE) {
```
use switch

vbnrh replied:
done

afreen23 commented on frontend/packages/ceph-storage-plugin/src/components/bucket-class/modals/edit-backingstore-modal.tsx:
```diff
@@ -54,34 +106,110 @@ const BucketClassEditModal = withHandlePromise<
-    bucketClass.spec.placementPolicy.tiers[0].backingStores = selectedA.map(getName);
-    if (policyB?.length) {
-      bucketClass.spec.placementPolicy.tiers[1].backingStores = selectedB.map(getName);
+    if (!isNamespaceType) {
+      bucketClass.spec.placementPolicy.tiers[0].backingStores = state.tier1BackingStore.map(
+        getName,
+      );
+      if (policyB?.length) {
+        bucketClass.spec.placementPolicy.tiers[1].backingStores = state.tier2BackingStore.map(
+          getName,
+        );
+      }
     }
+    if (isNamespaceType) {
```
one of them can be in else part.

vbnrh replied:
done

afreen23 commented on frontend/packages/ceph-storage-plugin/src/components/bucket-class/modals/edit-backingstore-modal.tsx:
```diff
@@ -99,19 +227,29 @@ const BucketClassEditModal = withHandlePromise<
-            setSelectedTierB={setSelectedB}
-            hideCreateBackingStore
-          />
+          {!isNamespaceType && (
+            <BackingStoreSelection
+              namespace={bucketClass.metadata.namespace}
+              tier1Policy={policyA}
+              tier2Policy={policyB}
+              selectedTierA={state.tier1BackingStore}
+              selectedTierB={state.tier2BackingStore}
+              setSelectedTierA={(selectedA) =>
+                dispatch({ type: 'setBackingStoreTier1', value: selectedA })
+              }
+              setSelectedTierB={(selectedB) =>
+                dispatch({ type: 'setBackingStoreTier2', value: selectedB })
+              }
+              hideCreateBackingStore
+            />
+          )}
+          {isNamespaceType && getNamespaceStorePage()}
```
you need conditional here
isNamespaceType ? ... : ...

vbnrh replied:
changed it. I am just used to lot of ifs than terinary conditions :)

vbnrh reviewed (commented):
> The namespace store is not present in my system and the page is breaking and that should not be the case. Need to have valid messaging.

![Screenshot from 2021-04-09 09-36-08](https://user-images.githubusercontent.com/57935785/114134627-aec64080-9925-11eb-852f-79e526c344a2.png)

I have fixed to show alerts. Please test and let me know @cloudbehl

afreen23 commented:
> > The namespace store is not present in my system and the page is breaking and that should not be the case. Need to have valid messaging.
> 
> ![Screenshot from 2021-04-09 09-36-08](https://user-images.githubusercontent.com/57935785/114134627-aec64080-9925-11eb-852f-79e526c344a2.png)
> 
> I have fixed to show alerts. Please test and let me know @cloudbehl

I think for 4.7 we want to hide selection of namespace store , if its not present. @vbnrh are we doing that ? if yes, then corresponding references would themselves hide.

afreen23 commented on frontend/packages/ceph-storage-plugin/src/components/bucket-class/modals/edit-backingstore-modal.tsx:
```diff
@@ -99,19 +228,30 @@ const BucketClassEditModal = withHandlePromise<
-              'ceph-storage-plugin~BackingStore represents a storage target to be used as the underlying storage for the data in Multicloud Object Gateway buckets.',
+              'ceph-storage-plugin~{{storeType}} represents a storage target to be used as the underlying storage for the data in Multicloud Object Gateway buckets.',
+              {
+                storeType: isNamespaceType ? 'NamespaceStore' : 'BackingStore',
```
externalize this

vbnrh replied:
done

afreen23 commented on frontend/packages/ceph-storage-plugin/locales/en/ceph-storage-plugin.json:
```diff
@@ -120,7 +121,7 @@
   "Create new BackingStore ": "Create new BackingStore ",
   "An error has occured while fetching backing stores": "An error has occured while fetching backing stores",
-  "Select a BackingStore": "Select a BackingStore",
+  "Select a backing store": "Select a backing store",
```
I think previous was okay.

vbnrh replied:
for namespace store it is like this. For consistency i have kept the backing store in same way.

afreen23 commented:
> > > The namespace store is not present in my system and the page is breaking and that should not be the case. Need to have valid messaging.
> > 
> > 
> > ![Screenshot from 2021-04-09 09-36-08](https://user-images.githubusercontent.com/57935785/114134627-aec64080-9925-11eb-852f-79e526c344a2.png)
> > I have fixed to show alerts. Please test and let me know @cloudbehl
> 
> I think for 4.7 we want to hide selection of namespace store , if its not present. @vbnrh are we doing that ? if yes, then corresponding references would themselves hide.

I think here you can set model feature flag for namespace store and then consume it using `useFlag` hook in your component.

vbnrh commented:
> > > > The namespace store is not present in my system and the page is breaking and that should not be the case. Need to have valid messaging.
> > > 
> > > 
> > > ![Screenshot from 2021-04-09 09-36-08](https://user-images.githubusercontent.com/57935785/114134627-aec64080-9925-11eb-852f-79e526c344a2.png)
> > > I have fixed to show alerts. Please test and let me know @cloudbehl
> > 
> > 
> > I think for 4.7 we want to hide selection of namespace store , if its not present. @vbnrh are we doing that ? if yes, then corresponding references would themselves hide.
> 
> I think here you can set model feature flag for namespace store and then consume it using `useFlag` hook in your component.

Thanks i've added feature flag for namespace-store. Please check.

afreen23 commented on frontend/packages/ceph-storage-plugin/src/components/bucket-class/modals/edit-backingstore-modal.tsx:
```diff
@@ -17,35 +17,87 @@ import {
 >((props) => {
   const { t } = useTranslation();
   const { bucketClass, inProgress, errorMessage, handlePromise, close, cancel } = props;
+  const isNamespaceType = !!bucketClass.spec?.namespacePolicy;
```
Also check for namespace store guard here if in case a policy can be added from CLI.
Other wise looks good.

vbnrh replied:
added the flag here too .. thanks

afreen23 commented:
/lgtm
